packages/twenty-server/src/engine/api/graphql/graphql-query-runner/resolvers/graphql-query-merge-many-resolver.service.ts

```diff
@@ -218,6 +243,238 @@ export class GraphqlQueryMergeManyResolverService extends GraphqlQueryBaseResolv
+    const fieldMetadata = Object.values(
+      objectMetadataItemWithFieldMaps.fieldsById,
+    ).find((field) => field?.name === fieldName);
+
+    if (!fieldMetadata || !isCompositeFieldMetadataType(fieldMetadata.type)) {
+      return null;
+    }
+
+    const compositeDefinition = compositeTypeDefinitions.get(
+      fieldMetadata.type,
+    );
+
+    if (!compositeDefinition) {
+      return null;
+    }
+
+    const mergedComposite: Record<string, unknown> = {};
+    const priorityRecord = recordsWithValues.find(
+      (item) => item.recordId === priorityRecordId,
+    );
+
+    compositeDefinition.properties.forEach((property) => {
+      const propertyName = property.name;
+
+      if (property.isIncludedInUniqueConstraint) {
+        mergedComposite[propertyName] = this.getUniqueConstraintValue(
+          priorityRecord,
+          propertyName,
+          recordsWithValues,
+        );
+
+        return;
+      }
+
+      if (property.isArray || property.type === FieldMetadataType.RAW_JSON) {
+        mergedComposite[propertyName] = this.mergeArrayProperty(
+          recordsWithValues,
+          propertyName,
+          compositeDefinition,
+        );
+
+        return;
+      }
+
+      mergedComposite[propertyName] = this.mergeSimplePropertyFromAllRecords(
+        priorityRecord,
+        recordsWithValues,
+        propertyName,
+      );
+    });
+
+    return mergedComposite;
+  }
+
+  private getUniqueConstraintValue(
+    priorityRecord: { value: unknown; recordId: string } | undefined,
+    propertyName: string,
+    recordsWithValues?: { value: unknown; recordId: string }[],
+  ): unknown {
+    const priorityValue = priorityRecord?.value as Record<
+      string,
+      unknown
+    > | null;
+
+    const priorityFieldValue = priorityValue?.[propertyName];
+
+    if (hasRecordFieldValue(priorityFieldValue)) {
+      return priorityFieldValue;
+    }
+
+    if (recordsWithValues) {
+      for (const record of recordsWithValues) {
+        const recordValue = record.value as Record<string, unknown> | null;
+        const fieldValue = recordValue?.[propertyName];
+
+        if (hasRecordFieldValue(fieldValue)) {
+          return fieldValue;
+        }
+      }
+    }
+
+    return null;
+  }
+
+  private mergeArrayProperty(
+    recordsWithValues: { value: unknown; recordId: string }[],
+    propertyName: string,
+    compositeDefinition: CompositeType,
+  ): unknown[] | null {
+    const allArrays: unknown[] = [];
+
+    recordsWithValues.forEach((record) => {
+      const recordValue = record.value as Record<string, unknown> | null;
+      const arrayValue = recordValue?.[propertyName];
+
+      if (Array.isArray(arrayValue)) {
+        allArrays.push(...arrayValue);
+      }
+    });
+
+    const uniqueArrays = this.deduplicateCompositeArrayItems(
+      allArrays,
+      compositeDefinition,
+    );
+
+    return uniqueArrays.length > 0 ? uniqueArrays : null;
+  }
+
+  private mergeSimplePropertyFromAllRecords(
```

Comment:
I'm a bit lost here to me these is a bit too abstract. Let's identify first what field metadata type we have (ARRAY, PHONES, ...) and write a dedicated function that would take an array of { recordId, fieldValue}[] (as you have) and merge them.
There is no need to abstract it further, it makes the code hard to read